dmriprep/workflows/dwi/util.py

```diff
@@ -0,0 +1,250 @@
+        validation_report
+            HTML reportlet indicating whether ``dwi_file`` had a valid affine
+
+
+    **Subworkflows**
+
+        * :py:func:`~dmriprep.workflows.dwi.util.init_enhance_and_skullstrip_wf`
+
+    """
+    workflow = Workflow(name=name)
+    workflow.__desc__ = """\
+First, a reference volume and its skull-stripped version were generated
+using a custom methodology taken from *fMRIPrep*.
+"""
+    inputnode = pe.Node(niu.IdentityInterface(fields=['dwi_file', 'bvec_file', 'bval_file']),
+                        name='inputnode')
+    outputnode = pe.Node(
+        niu.IdentityInterface(fields=['dwi_file', 'raw_ref_image',
+                                      'ref_image', 'ref_image_brain',
+                                      'dwi_mask', 'validation_report']),
+        name='outputnode')
+
+    # Simplify manually setting input image
+    if dwi_file is not None:
+        inputnode.inputs.dwi_file = dwi_file
+
+    validate = pe.Node(ValidateImage(), name='validate', mem_gb=DEFAULT_MEMORY_MIN_GB)
+
+    gen_ref = pe.Node(MeanB0(), name="gen_ref")
+
+    pre_mask = pe.Node(afni.Automask(outputtype="NIFTI_GZ"), name="pre_mask")
+
+    enhance_and_skullstrip_dwi_wf = init_enhance_and_skullstrip_dwi_wf(
+        omp_nthreads=omp_nthreads)
+
+    workflow.connect([
+        (inputnode, validate, [('dwi_file', 'in_file')]),
+        (validate, gen_ref, [('out_file', 'in_file')]),
+        (inputnode, gen_ref, [('bvec_file', 'in_bvec'),
+                              ('bval_file', 'in_bval')]),
+        (gen_ref, pre_mask, [('out_file', 'in_file')]),
+        (gen_ref, enhance_and_skullstrip_dwi_wf, [('out_file', 'inputnode.in_file')]),
+        (pre_mask, enhance_and_skullstrip_dwi_wf, [('out_file', 'inputnode.pre_mask')]),
+        (validate, outputnode, [('out_file', 'dwi_file'),
+                                ('out_report', 'validation_report')]),
+        (gen_ref, outputnode, [('out_file', 'raw_ref_image')]),
+        (enhance_and_skullstrip_dwi_wf, outputnode, [
+            ('outputnode.bias_corrected_file', 'ref_image'),
+            ('outputnode.mask_file', 'dwi_mask'),
+            ('outputnode.skull_stripped_file', 'ref_image_brain')]),
+    ])
+
+    if gen_report:
+        mask_reportlet = pe.Node(SimpleShowMaskRPT(), name='mask_reportlet')
+        workflow.connect([
+            (enhance_and_skullstrip_dwi_wf, mask_reportlet, [
+                ('outputnode.bias_corrected_file', 'background_file'),
+                ('outputnode.mask_file', 'mask_file'),
+            ]),
+        ])
+
+    return workflow
+
+
+def init_enhance_and_skullstrip_dwi_wf(
+        name='enhance_and_skullstrip_dwi_wf',
+        omp_nthreads=1):
+    """
+    This workflow takes in a dwi reference iamge and sharpens the histogram
+    with the application of the N4 algorithm for removing the
+    :abbr:`INU (intensity non-uniformity)` bias field and calculates a signal
+    mask.
+
+    Steps of this workflow are:
```

Comment:
I've been thinking about this, and I would like to hear opinions (@arokem, @jelleveraart) about a more robust b0 computation:

Using the rough mask (I'll call it `pre_mask`) calculated before, what about recalculating the reference as follows:

```Python
bzeros = dwi_data[..., b0vals_mask]
bzeros = 1000 * bzeros / bzeros[pre_mask, ...].mean(axis=-1)  # Normalize to have mean intensity of 1000 within mask
b0map = np.median(bzeros, axis=-1)  # median should be more robust than average, esp. with few bzeros
```

Reply:
The first rescaling is to deal with the average signal decay you generally see when b zeros are distributed along the scan. We could also store the average signal series (that `bzeros[pre_mask, ...].mean(axis=-1)`) to have an estimation of signal drop and maybe allow some exponential decay fitting down the line.

We could also robustify that calculation using median instead of mean.

Reply:
Would it also be worth registering each b0 to the first in case there is significant motion between them? I'm not sure what opinions are on this.

Reply:
Yes. I would start by registering all the b0 to each other, and then registering the others to the mean of these registered b0 volumes.

Reply:
I've temporarily gotten FSL mcflirt setup but would this be something handled by antsMultivariateTemplateConstruction?